src/_pytest/main.py

```diff
@@ -540,10 +547,22 @@ def pytest_runtest_logreport(
     def isinitpath(self, path: Union[str, "os.PathLike[str]"]) -> bool:
         return Path(path) in self._initialpaths
 
-    def gethookproxy(self, fspath: "os.PathLike[str]"):
+    def gethookproxy(self, fspath: "os.PathLike[str]") -> PathAwareHookProxy:
         # Check if we have the common case of running
         # hooks with all conftest.py files.
+
         pm = self.config.pluginmanager
+        len_conftest_plugins = len(pm._conftest_plugins)
+
+        if len_conftest_plugins != self._num_conftest_plugins_collected:
```

Comment:
a possible follow-up could be to hook into plugin registration and only discarding affected hook proxies

Reply:
makes more sense than using the `_num_conftest_plugins_collected` variable. Removed this variable and cache cleanup is now done inside `pytest_plugin_registered`